packages/sdk/src/Actions/EventAction.ts

```diff
@@ -508,26 +543,92 @@ export class EventAction extends DeployableTarget<
     }
     return true;
   }
-
   /**
-   * Validates a {@link Log} against a given criteria.
+   * Validates a single action function with a given criteria against logs.
+   * If logs are provided in the optional `params` argument, then those logs will be used instead of fetched with the configured client.
    *
-   * @param {Criteria} criteria - The criteria to validate against.
-   * @param {Log} log - The Viem event log.
-   * @returns {boolean} - Returns true if the log passes the criteria, false otherwise.
+   * @public
+   * @async
+   * @param {ActionStep} actionStep
+   * @param {?GetLogsParams<Abi, ContractEventName<Abi>> & {
+   *       knownEvents?: Record<Hex, AbiEvent>;
+   *       txHash?: Hex;
+   *     }} [params]
+   * @returns {Promise<boolean>}
    */
-  public validateLogAgainstCriteria(criteria: Criteria, log: Log) {
-    const fieldValue = log.topics.at(criteria.fieldIndex);
-    if (fieldValue === undefined) {
-      throw new FieldValueUndefinedError({ log, criteria, fieldValue });
+  public async isActionFunctionValid(
+    actionStep: ActionStep,
+    params?: GetLogsParams<Abi, ContractEventName<Abi>> & {
+      knownEvents?: Record<Hex, AbiEvent>;
+      txHash?: Hex;
+    },
+  ) {
+    const criteria = actionStep.actionParameter;
+    const signature = trim(actionStep.signature);
+    if (!params || !params?.txHash) {
+      // Should we return false in this case?
+      throw new Error('txHash is required for function validation');
     }
+    const client = this._config.getClient({
+      chainId: params?.chainId,
+    }) as PublicClient;
+    // Fetch the transaction receipt and decode the function input using `viem` utilities
+    const transaction = await client.getTransaction({ hash: params.txHash });
+    const func = (functions.abi as Record<Hex, AbiFunction>)[
+      signature
+    ] as AbiFunction;
+
+    if (!func) {
+      throw new Error(
+        `No known ABI for given function signature: ${signature}`,
+      );
+    }
+    let decodedData;
+    try {
+      decodedData = decodeFunctionData({
+        abi: [func],
+        data: transaction.input,
+      });
+    } catch (e) {
+      throw new Error(`Failed to decode function data: ${e}`);
```

Comment:
Let's type this error in `errors.ts`, giving it the input and abi as metadata

Reply:
You want me to type the other errors? There are other untyped errors in play here.